### PR TITLE
add Prometheus metrics endpoint

### PR DESCRIPTION
**Pull request by hawkw**
This branch builds on PR #107 to add a production-ready Prometheus
metrics endpoint. As discussed in #107, the metrics endpoint is
implemented using the `metrics-endpoint-prometheus` and
`metrics-process` crates, and will serve process metrics by default. If
the application or other libraries use the `metrics` crate to emit their
own metrics, these metrics will be included in the Prometheus endpoint
as well.

Changes from #107 include the following:

- The metrics endpoint is feature-flagged behind the "metrics" feature.
- Metrics are compressed using GZIP when the request includes an
  `accept-encoding: gzip` header
- The `PrometheusBuilder` type from `metrics-exporter-prometheus` is
  wrapped in a newtype. This is because `metrics-exporter-prometheus`
  includes functionality for configuring an allowlist of allowed IP
  networks, which will deny any requests not sent by one of those IP
  addresses. If this is configured using
  `metrics_exporter_prometheus::PrometheusBuilder` and then the builder
  is passed to `kubert`, however, the allowlist will not work.
  Therefore, we can wrap the builder in our own type that passes through
  most of the methods and re-implements the allowlist functionality. In
  addition, other `metrics-*` types used by kubert are re-exported from
  the `admin::metrics` module so that they can be used without requiring
  an import of the various `metrics-*` crates in the end-user project.

Closes #107
Closes adleong/kubert#1

### CODE REVIEW

**Comment by hawkw** [comment]
One potential question is whether the IP address allowlist should just be implemented for every route on the admin server, instead.

**Comment by hawkw** [comment]
Bleh, this adds a bunch of `cargo-deny` failures due to transitive deps of `metrics-exporter-prometheus`: https://github.com/olix0r/kubert/actions/runs/4001001544/jobs/6866778473

Is it okay to add exceptions for these, or will we need to make changes upstream to fix the duplicate versions?

**Review by adleong — COMMENTED** [review]
Sweet!

I wonder if wrapping the PrometheusBuilder with our own type and forwarding methods is really worth it in order to preserve the allowlist functionality.  Especially since this type of functionality can be better implemented using Linkerd authorization policy.

Should we perhaps just instead document that any PrometheusBuilder allowlist settings will not be respected by kubert?

**Comment by hawkw** [comment]
> I wonder if wrapping the PrometheusBuilder with our own type and forwarding methods is really worth it in order to preserve the allowlist functionality. Especially since this type of functionality can be better implemented using Linkerd authorization policy.

I could be convinced either way --- my initial plan was to wrap the builder just to _remove_ the allowlist functions, but I realized that it was quite simple to implement the allowlist. However, I'd be fine with removing that functionality and documenting that those methods on the `metrics-exporter-prometheus` type will be ignored.

I do think that there's kind of a side benefit for having our own type or re-exporting the builder, which is that user code doesn't need to have its own dependency on `metrics-exporter-prometheus`. This means that the same version of `metrics-exporter-prometheus` that `kubert` uses will always be used by the user code that configures the builder, so there aren't version mismatch issues.

**Review by olix0r — CHANGES_REQUESTED** [review]
> I wonder if wrapping the PrometheusBuilder with our own type and forwarding methods is really worth it in order to preserve the allowlist functionality.

Yeah. I'd prefer to avoid having to basically reimplement their builder type.

How would allowlists be used in the context of Kubernetes?

Can we simplify this a lot of we remove this requirement? What about something like:

* Add to the admin server a map of additional, synchronous handlers. Basically something like an `Arc<dyn Fn(http::Request<hyper::Body>) -> http::Response<hyper::Body>>`. We add a fn on the builder to add arbitrary admin endpoints.
* In the metrics module have some function that takes a `PrometheusHandle` and returns a handler.
* Together, this could support multiple metrics endpoints from a single admin server, right?
* Probably add some other util functions to help initialization in the metrics module and perhaps a builder bundle like `with_default_prometheus()` or something.

**Comment by hawkw** [comment]
Okay, I'm happy to remove the allowlist stuff!

**Review by adleong — APPROVED** [review]

**Comment by olix0r** [comment]
I did some spelunking in indexmap & hashbrown re: duplicate versions of `ahash`. hashbrown v0.13 bumps the MSRV to v1.61, so I suspect we won't get a backport fix for indexmap v1.9.x. hashbrown v0.13 is available in `indexmap v2.x.x`, but that's still in pre-release.

I think we'll need to allow the exception (with a comment). cargo-deny needs some other licensing acknowledgements as well.

File: kubert/src/admin.rs [hunk 1]
```diff
@@ -29,19 +37,19 @@ pub struct AdminArgs {
...
 }
 
 /// Supports spawning an admin server
 #[cfg_attr(docsrs, doc(cfg(feature = "admin")))]
-#[derive(Debug)]
 pub struct Bound {
     addr: SocketAddr,
     ready: Readiness,
     server: hyper::server::Builder<hyper::server::conn::AddrIncoming>,
+    routes: HashMap<String, HandlerFn>,
```
Comment (olix0r) [inline kubert/src/admin.rs:52]: We already have an `ahash` dependency... so might as well use it here.

**Comment by hawkw** [comment]
cool, i'll add the allows then!

**Comment by olix0r** [comment]
Some small suggestions here https://github.com/hawkw/kubert/pull/1

**Review by olix0r — APPROVED** [review]

**Pull request merged**